File: 2021/day2/day2.hs

toTuple :: String -> (String,Int)
toTuple x = (head y, read (y !! 1))
    where
        y = words x

part1 :: [(String,Int)] -> (Int,Int)
part1 (("forward", x):xs) = (fst (part1 xs), snd (part1 xs) + x)
part1 (("up", x):xs) = (fst (part1 xs) - x, snd (part1 xs))
part1 (("down", x):xs) = (fst (part1 xs) + x, snd (part1 xs))
part1 _ = (0,0)

part2 :: [(String,Int)] -> Int -> (Int,Int)
part2 (("forward", x):xs) aim = (depth + x * aim, forward + x)
    where (depth, forward) = part2 xs aim
part2 (("up", x):xs) aim = part2 xs (aim - x)
part2 (("down", x):xs) aim = part2 xs (aim + x)
part2 _ _ = (0,0)

main :: IO ()
main = do
    inData <- readFile "input.txt"
    let input = map toTuple (lines inData)
    print $ uncurry (*) $ part1 input
    print $ uncurry (*) $ part2 input 0

   

    
    
    
            

File: 2021/day2/day2.py


with open("./input.txt") as f:
    lines = f.readlines()

    data = []
    for z in lines:
        y = z.split()
        data.append((y[0], int(y[1])))

    x = 0
    depth = 0
    aim = 0
    for inst in data:
        if inst[0] == "forward":
            x += inst[1]
            depth += inst[1] * aim
        elif inst[0] == "down":
            aim += inst[1]
        elif inst[0] == "up":
            aim -= inst[1]

    print(x*aim)
    print(x*depth)